src/components/action-popover/action-popover-item/action-popover-item.component.tsx

```diff
@@ -124,42 +133,96 @@ export const ActionPopoverItem = ({
     "ActionPopoverItem only accepts submenu of type `ActionPopoverMenu`"
   );
 
-  const { setOpenPopover, isOpenPopover, focusButton } = context;
+  const {
+    setOpenPopover,
+    isOpenPopover,
+    focusButton,
+    submenuPosition,
+  } = context;
   const isHref = !!href;
   const [containerPosition, setContainerPosition] = useState<
     ContainerPosition | undefined
   >(undefined);
   const [guid] = useState(createGuid());
   const [isOpen, setOpen] = useState(false);
   const [focusIndex, setFocusIndex] = useState<number>(0);
-  const [isLeftAligned, setIsLeftAligned] = useState(true);
 
   const submenuRef = useRef<HTMLDivElement>(null);
   const ref = useRef<HTMLButtonElement>(null);
   const mouseEnterTimer = useRef<NodeJS.Timeout | null>(null);
   const mouseLeaveTimer = useRef<NodeJS.Timeout | null>(null);
+  const hasSubmenuAligned = useRef(false);
 
   useEffect(() => {
     if (!isOpenPopover) {
       setOpen(false);
     }
   }, [isOpenPopover]);
 
+  useEffect(() => {
+    if (icon) {
+      setChildHasIcon?.(true);
+    }
+    if (submenu) {
+      setChildHasSubmenu?.(true);
+    }
+  }, [icon, setChildHasSubmenu, setChildHasIcon, submenu]);
+
   const alignSubmenu = useCallback(() => {
-    if (checkRef(ref) && checkRef(submenuRef) && submenu) {
-      const align = leftAlignSubmenu(ref, submenuRef);
-      setIsLeftAligned(align);
-      setContainerPosition(getContainerPosition(ref, submenuRef, placement));
+    const checkCalculatedSubmenuPosition = calculateSubmenuPosition(
+      ref,
+      submenuRef,
+      submenuPosition,
+      currentSubmenuPosition
+    );
+
+    setCurrentSubmenuPosition?.(checkCalculatedSubmenuPosition);
+
+    return checkRef(ref) && checkRef(submenuRef) && submenu;
+  }, [
+    submenu,
+    setCurrentSubmenuPosition,
+    submenuPosition,
+    currentSubmenuPosition,
+  ]);
+
+  useEffect(() => {
+    function getContainerPosition(
```

Comment:
**comment** (non-blocking): I see that this function, which previously was defined outside the scope of the component, is now inside (and correctly inside the `useEffect` since it's not called from anywhere else). It doesn't matter much but this could be moved back outside - while it does use values from the component scope, this could still work with the function defined outside by passing those values in as additional arguments.

I'm not saying this would be better - 6 arguments would be rather a lot! - but it seems inconsistent to pass in 3 values as arguments but not the other 3. I'd prefer to either have the function defined outside the component and pass all 6 in as arguments, or have it here and read everything from the component scope (and add them as effect dependencies).

Not important though so don't feel under any obligation.

Reply:
Yeah I agree with Robin, I don't think we need to pass any arguments here as it will have access to the refs/props